### ISSUE TITLE
Helper classes

### ISSUE THREAD

**Issue by dedmen** [intercept/intercept#110]
I can't seem to make up a better title for this. Just see yourself:
```
class Variable {
public:
    Variable() {};
    Variable(std::string varName) : _name(varName), _namespace(intercept::sqf::mission_namespace()) {};//#TODO namespace/object and other types constructor argument
    auto get() {
        return intercept::sqf::get_variable(_namespace, _name);
    }
    auto set(const game_value &gv, bool publicVar = false) {
        intercept::sqf::set_variable(_namespace, _name, gv);
    }
	/*
    auto getCache(maxTime){
    if (now - lastCache < maxTime)
		return _cached;
	else
		return get();
    }
    */
private:
    rv_namespace _namespace;
    std::string _name;
};
```

in code:

```
void __cdecl intercept::on_frame() {
Variable myVar("someGlobalVar");

intercept::sqf::side_chat(myVar.get(), "hello");
};
```
Same for object/location/marker and such. Combining all functions that are relative to that object in one class. Object for example can have a constructor that automatically spawns the object and would despawn it on destruct.

**Comment by Glowbal** [comment]
yeah, this is a good improvement once the core functionality is done. We should then implement a sensible library on top of the sqf wrapper.

**Comment by dedmen** [comment]
As I see right now the Variable function also allows for this
```c++
someGlobalVar = Variable("someGlobalVar") = intercept::sqf::player();
```
which will assign a value to a SQF and a C++ Variable.

Feel free to post more examples for such helpers so we have enough Idea's when we will implement that someday.

**Comment by dedmen** [comment]
ppEffectAdjust parameter is dependent on the type of ppEffect it's used on. So we should probably just make a templated helper class for ppEffects and specialize it for each effect type with the adjust function taking different parameters for each.

**Comment by chris579** [comment]
I would also go for a player object that makes it easier to interact with players.

Pseudocode

```
class Player {
public:
    Player(intercept::types::object player) {
      _containingObject = player;
    };

    void setPos(intercept::types::vector3 pos) {
      intercept::sqf::setPos(_containingObject, pos);
    }

    auto getPos() {
      return intercept::sqf::getPos(_containingObject);
    }

    auto isAlive() {
      return intercept::sqf::alive(_containingObject);
    }

    // and so on    

private:
    intercept::types::object player _containingObject;
};
```

Therefore `intercept::sqf::player` would return a player instance.

This could be extended with static methods for getting players by name or by uid and so on.

**Comment by dedmen** [comment]
That would be a Unit/Man class. As set_pos and stuff are not unique to Player itself. The player is only the Unit that is currently possessed by the player.

**Comment by chris579** [comment]
I would suggest inheritance for this case.
`Man -> Unit -> Player`

or a field indicating that the unit is the current player or a player in general

**Comment by dedmen** [comment]
```c++
class Player: public Unit {
public:
    Player() : Unit(sqf::player()) {}
}
```
I'll try to get some UML going.

**Comment by Tirpitz93** [comment]
Regarding your UML shouldn't `Person` be `Man`? AFAIK that's how this type of object is referred to in SQF etc.

**Comment by dedmen** [comment]
What about women? Or dogs. or UAV controllers. Okey a dog also isn't really a Person.
Man is a config class that is commonly used as a parent for a human looking AI unit yeah.

**Comment by Tirpitz93** [comment]
@dedmen so for the purposes of this exercise wouldn't it make sense for it to be man, mirroring the config class?
And what is the inheritance of `group` objects. do they inherit from object or are they have a seperate inheritance tree?

**Comment by killerswin2** [comment]
This a good example of a [proxy pattern](https://en.wikipedia.org/wiki/Proxy_pattern)
